libs/platform/ebpf_hash_table.c

```diff
@@ -3,52 +3,175 @@
 
 #include "ebpf_platform.h"
 
+typedef struct _ebpf_hash_bucket_header
+{
+    struct _ebpf_hash_bucket_header* next;
+    uint8_t data[1];
+} ebpf_hash_bucket_header_t;
+
 struct _ebpf_hash_table
 {
-    struct _RTL_AVL_TABLE avl_table;
-    ebpf_hash_table_compare_result_t (*compare_function)(const uint8_t* key1, const uint8_t* key2);
+    uint32_t bucket_count;
+    uint32_t entry_count;
+    uint32_t seed;
     size_t key_size;
     size_t value_size;
     void* (*allocate)(size_t size);
     void (*free)(void* memory);
+    void (*extract)(const uint8_t* value, const uint8_t** data, size_t* num);
+    ebpf_hash_bucket_header_t* buckets[1];
 };
 
-// NOTE:
-// AVL tree gives a single struct containing both key and value.
-// Compare can be called with a partial struct only containing the key.
-// Do not access beyond map->ebpf_map_definition.key_size bytes.
-static RTL_GENERIC_COMPARE_RESULTS
-_ebpf_hash_map_compare(_In_ struct _RTL_AVL_TABLE* avl_table, _In_ void* first_struct, _In_ void* second_struct)
+/**
+ * @brief Perform a rotate left on a value.
+ *
+ * @param value Value to be rotated.
+ * @param count Count of bits to rotate.
+ * @return Rotated value.
+ */
+static inline unsigned long
+_ebpf_rol(uint32_t value, size_t count)
+{
+    return (value << count) | (value >> (32 - count));
+}
+
+// Ported from https://github.com/aappleby/smhasher
```

Comment:
That repo has no license listed.  What license is this code based on?

Reply:
It is a re-implementation based on the algorithm:
https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash3.cpp#L2

License says public domain.

Reply:
Great, would be nice to say so in the // comment.